File: app/src/main/java/io/github/zwieback/familyfinance/core/lifecycle/destroyer/EntityViewDestroyer.java
package io.github.zwieback.familyfinance.core.lifecycle.destroyer;

import androidx.annotation.NonNull;

import java.sql.Connection;
import java.sql.Statement;
import java.util.concurrent.Callable;

import io.reactivex.Single;

public abstract class EntityViewDestroyer implements Callable<Single<Boolean>> {

    private final Connection connection;

    protected EntityViewDestroyer(Connection connection) {
        this.connection = connection;
    }

    @NonNull
    protected abstract String getViewName();

    /**
     * @return {@code false} because "DROP VIEW" is DDL operation
     */
    @Override
    public Single<Boolean> call() {
        return Single.fromCallable(() -> {
            try (Statement statement = connection.createStatement()) {
                return statement.execute("DROP VIEW IF EXISTS " + getViewName());
            }
        });
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/core/lifecycle/destroyer/EntityViewDestroyer.kt
package io.github.zwieback.familyfinance.core.lifecycle.destroyer

import io.reactivex.Single
import java.sql.Connection
import java.util.concurrent.Callable

abstract class EntityViewDestroyer protected constructor(
    private val connection: Connection
) : Callable<Single<Boolean>> {

    protected abstract val viewName: String

    /**
     * @return `false` because "DROP VIEW" is DDL operation
     */
    override fun call(): Single<Boolean> {
        return Single.fromCallable {
            connection.createStatement()
                .use { statement -> statement.execute("DROP VIEW IF EXISTS $viewName") }
        }
    }
}
